speech-commands/src/browser_fft_recognizer.ts

```diff
@@ -754,30 +767,70 @@ class TransferBrowserFftSpeechCommandRecognizer extends
       this.createTransferModelFromBaseModel();
     }
 
+    // This layer needs to be frozen for the initial phase of the
+    // transfer learning. During subsequent fine-tuning (if any), it will
+    // be unfrozen.
+    this.secondLastBaseDenseLayer.trainable = false;
+
     // Compile model for training.
-    const optimizer = config.optimizer || 'sgd';
-    this.model.compile(
-        {loss: 'categoricalCrossentropy', optimizer, metrics: ['acc']});
+    this.model.compile({
+      loss: 'categoricalCrossentropy',
+      optimizer: config.optimizer || 'sgd',
+      metrics: ['acc']
+    });
 
     // Prepare the data.
     const {xs, ys} = this.collectTransferDataAsTensors();
 
-    const epochs = config.epochs == null ? 20 : config.epochs;
-    const validationSplit =
-        config.validationSplit == null ? 0 : config.validationSplit;
+    let trainXs: tf.Tensor;
+    let trainYs: tf.Tensor;
+    let valData: [tf.Tensor, tf.Tensor];
     try {
-      const history = await this.model.fit(xs, ys, {
-        epochs,
-        validationSplit,
+      if (config.validationSplit != null) {
+        const splits = balancedTrainValSplit(xs, ys, config.validationSplit);
+        trainXs = splits.trainXs;
+        trainYs = splits.trainYs;
+        valData = [splits.valXs, splits.valYs];
+      } else {
+        trainXs = xs;
+        trainYs = ys;
+      }
+
+      const history = await this.model.fit(trainXs, trainYs, {
+        epochs: config.epochs == null ? 20 : config.epochs,
+        validationData: valData,
         batchSize: config.batchSize,
         callbacks: config.callback == null ? null : [config.callback]
       });
-      tf.dispose([xs, ys]);
-      return history;
-    } catch (err) {
-      tf.dispose([xs, ys]);
-      this.model = null;
-      return null;
+
+      if (config.fineTuningEpochs != null && config.fineTuningEpochs > 0) {
```

Comment:
is fineTuning phase an example for the chapter or it provides better result if it is done separately after the transfer learning?
And should the layer trainable get reset after the fine tuning is done?